docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,420 @@
+  - A plugin might require some setup or teardown when it is registered or
+    unregistered, so these functions are required. Furthermore, a plugin
+    might require configuration, which is why there is a `config_data`
+    parameter. The configuration data can be plugin specific, so no format is
+    specified in this proposal.
+- `get_evidence` and `free_evidence`
+  - Producing evidence is necessary for attestation.
+  - There is a `custom_claims` parameter because many attestation protocols
+    require the enclave to sign some claim from a relying party. For example,
+    many protocols follow the "challenge response" architecture, which requires
+    the enclave to sign a nonce from the relying party.
+- `get_collateral` and `free_collateral`
+  - Producing collateral is essential for attestation.
+  - Examples of collateral could be firmware measurements from the device's
+    manufacturer or CRLs from an X509 certificate.
+- `verify_evidence`
+  - Verifying evidence and collateral is essential for attestation.
+  - The `claims` field contains key-value pairs that can be verified by the
+    caller. This will have the similar contents as the `oe_identity_t` field
+    in the `oe_report_t` struct and any custom claims that were passed to the
+    `get_evidence` function.
+  - Should at minimum have the following claims (based of `oe_identity_t`):
+    - `id_version`: Version of the OE claims. Must be 0.
+    - `security_version`: Security version of the enclave. (ISVN for SGX).
+    - `attributes`: Values of the attributes flags for the enclave:
+      - `OE_REPORT_ATTRIBUTES_DEBUG`: The report is for a debug enclave.
+      - `OE_REPORT_ATTRIBUTES_REMOTE`: The report can be used for remote
+        attestation.
+    - `unique_id`: The unique ID for the enclave (MRENCLAVE for SGX).
+    - `signer_id`: The signer ID for the enclave (MRSIGNER for SGX).
+    - `product_id`: The product ID for the enclave (ISVPRODID for SGX).
+
+TODO:
+
+- Format of claims. The format should be independent of the plugin. It also
+  should be easy to extract the claims to chain plugins. For example, an app
+  might use the SGX plugin to verify the SGX quote and get all the claims,
+  including the custom ones. Then, it can something like a JWT token plugn to
+  write all of those claims as evidence. If the format is something like JSON
+  or CBOR, then this could work.
+- Input parameters for `get_evidence`, `get_collateral`, `verify_evidence`.
+  For `get_evidence`, there could pontentially be 3 types of input: 1) input
+  to the function itself 2) custom claims that are known to the plugin 3)
+  custom clams that are unknown to the plugin and should be treated as a
+  opaque block. The current API proposal has no way to distinguish all 3.
+  Likewise, `get_collateral` and `verify_evidence` could require function
+  input parameters. A solution could be having a structured way to define
+  these parameters using JSON/CBOR.
+
```

Comment:
What's your guys' preferred solution?

Reply:
I think we need to define a common header first for this input parameters.  The structure of the data will up to each plug-in, as long as the data is in a single blob of memory.

Reply:
Looking at Qve code for SGX, the collateral struct is almost all char strings while the evidence (the sgx_quote) is all binary.

Reply:
I added a section about the header.